Uchu.Core/Logging/LogQueue.cs

```diff
@@ -0,0 +1,135 @@
+using System;
+using System.Threading;
+
+namespace Uchu.Core
+{
+    public class LogEntry
+    {
+        public string message { get; set; }
+        public ConsoleColor color { get; set; }
+        public LogEntry nextEntry { get; set; }
+    }
+    
+    public class LogQueue : IDisposable
+    {
+        private readonly Mutex logMutex = new Mutex();
+        private LogEntry nextEntry = null;
+        private LogEntry lastEntry = null;
+        
+        /// <summary>
+        /// Adds a line to the log queue.
+        /// </summary>
+        public void AddLine(string line, ConsoleColor color)
+        {
+            // Lock the log.
+            logMutex.WaitOne();
+            
+            // Create the new entry.
+            var entry = new LogEntry()
+            {
+                message = line,
+                color = color,
+            };
+            
+            // Update the pointers.
+            if (this.nextEntry == null)
+            {
+                this.nextEntry = entry;
+                this.lastEntry = entry;
+            }
+            else
+            {
+                this.lastEntry.nextEntry = entry;
+                this.lastEntry = entry;
+            }
+
+            // Unlock the log.
+            this.logMutex.ReleaseMutex();
+        }
+
+        /// <summary>
+        /// Pops the next entry to output.
+        /// Returns null if there is no pending entry.
+        /// </summary>
+        public LogEntry PopEntry()
+        {
+            // Lock the log.
+            logMutex.WaitOne();
+            
+            // Pop the next entry.
+            LogEntry entry = null;
+            if (this.nextEntry != null)
+            {
+                entry = this.nextEntry;
+                this.nextEntry = entry.nextEntry;
+            }
+            
+            // Unlock the log.
+            this.logMutex.ReleaseMutex();
+            
+            // Return the entry.
+            return entry;
+        }
+
+        /// <summary>
+        /// Starts a thread for outputting the log. This prevents
+        /// the console IO from blocking the application.
+        /// </summary>
+        public void StartConsoleOutput()
+        {
+            new Thread(() =>
```

Comment:
Why wouldn't you just Task.Run this? Then the ThreadPoolScheduler will determine the thread that you get, but it'll still run in the background.

Reply:
I did this initially because of concerns of blocking the program by `Console.WriteLine` blocking due to Window's console window system. However, using `Console.Out.WriteLineAsync` does not have this problem and a `Task` can be used for what seems to have little to no effect. The change has been made.